c++/src/kj/mutex.h

```diff
@@ -172,6 +220,44 @@ class Mutex {
   mutable pthread_rwlock_t mutex;
 #endif
 
+#if KJ_SAVE_ACQUIRED_LOCK_INFO
+  uint32_t lockedExclusivelyByThread = 0;
+  SourceLocation lockAcquiredLocation;
+
+  volatile bool updatedSharedMetadata = false;
+
+  void acquiredExclusive(uint32_t tid, const SourceLocation& location) noexcept {
+    __atomic_store_n(&lockedExclusivelyByThread, tid, __ATOMIC_RELEASE);
+    lockAcquiredLocation = location;
```

Comment:
TLDR: Yes in theory, but how? Do I use `__atomic_thread_fence` with `__ATOMIC_RELAXED`? Assuming yes, my reading of the GCC man page would indicate that `__atomic_thread_fence` only impacts memory that was mutated by atomics per the C11 memory model (which I think is consistent with C++, right?).

This is a multi-word composite object so there's not really any atomics that would work (& this must be a copy unlike the blocking TLS metadata since this is "global" for the instance of the mutex). I think that's why I had non-relaxed semantics to try to control the visibility but that probably wouldn't have worked anyway as written (probably would need to carefully maintain a sentinel value).

I had actually thought about the line of reasoning you expressed above:

> because we don't start trying to trace locks until they are held for a long time

but my unease with that line of thinking was that the watchdog is racing ongoing lock/unlock operations so there is actually good reason to want this to be consistent & timely or we'll end up with superfluous Sentry issues (due to fingerprinting separating it from the underlying issue) that will also be missing critical data to make it easy to triage (or mislead the triager down the wrong analysis path). That being said, I think I can convince myself that this is a marginal scenario not worth worrying about for now.

Reply:
I suppose you could use `tid` to protect it:

* In this function, write `tid` _after_ `lockAcquiredLocation`, with `__ATOMIC_RELEASE`. This ensures that `lockAquiredLocation` is fully written before `tid`.
* When releasing, set `tid` back to zero _before_ modifying `lockAcquiredLocation`, with `__ATOMIC_ACQUIRE`. This ensures the subsequent modification of `lockAcquiredLocation` happens only after `tid` is already zero.
* In `lockedInfo()`, read `tid` with `__ATOMIC_ACQUIRE` before accessing `lockAcquiredLocation`. If `tid` is zero, assume `lockAcquiredLocation` is invalid.

I guess in `lockedInfo()` you currently assume `tid` of zero means a shared lock, but that assumption seems sketchy -- `tid` might be zero just because the locking thread is already in the process of unlocking.

I'm not sure if there's a good way to extend this to shared locks, though.

Admittedly this protection doesn't accomplish very much. But if we're not really rigorously protecting `location` then I would say there's no point is using atomics on `tid` either.

Reply:
I've rewritten the `lockedInfo` function to just read atomically from state instead of trying to guess at it from variables that are impossible to synchronize. At least this way while the data may be read stale, at least we'll get the right type of lock.